xdsl/interpreter.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from io import StringIO
+from typing import Any, Callable, Generator, TypeVar
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import SSAValue, Operation
+
+_OperationInvT = TypeVar('_OperationInvT', bound=Operation)
+
+
+class IntepretationError(Exception):
+    pass
+
+
+@dataclass
+class Functions:
+    functions: dict[type[Operation],
+                    Callable[[Intepreter, Operation, tuple[Any, ...]],
+                             tuple[Any, ...]]] = field(default_factory=dict)
+
+    def register_op(
+        self, op_type: type[_OperationInvT],
+        func: Callable[[Intepreter, _OperationInvT, tuple[Any, ...]],
+                       tuple[Any, ...]]):
+        self.functions[op_type] = func  # type: ignore
+
+    def op_types(self) -> set[type[Operation]]:
+        return set(self.functions.keys())
+
+    def register(
+        self, op_type: type[_OperationInvT]
+    ) -> Callable[[
+            Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                          ...]]
+    ], Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                     ...]]]:
+
+        def wrapper(
+            func: Callable[[Intepreter, _OperationInvT, tuple[Any, ...]],
+                           tuple[Any, ...]]):
+            self.register_op(op_type, func)
+            return func
+
+        return wrapper
+
+    def run(self, interpreter: Intepreter, op: Operation,
+            args: tuple[Any, ...]) -> tuple[Any, ...]:
+        return self.functions[type(op)](interpreter, op, args)
+
+    def register_from(self, other: Functions):
+        '''If there are duplicate definitions, the `other` will override `self`'''
+        self.functions.update(other.functions)
+
+
+@dataclass
+class IntepretationEnv:
+    name: str = field(default="unknown")
+    parent: IntepretationEnv | None = None
+    env: dict[SSAValue, Any] = field(default_factory=dict)
+
+    def __getitem__(self, key: SSAValue) -> SSAValue:
+        if key in self.env:
+            return self.env[key]
+        if self.parent is not None:
+            return self.parent[key]
+        raise IntepretationError(f'Could not find value for {key}')
+
+    def __setitem__(self, key: SSAValue, value: Any):
+        if key in self.env:
+            raise IntepretationError(
+                f'Attempting to register SSAValue {value} for name {key}'
+                ', but value with that name already exists')
+        self.env[key] = value
+
+    def stack(self) -> Generator[IntepretationEnv, None, None]:
```

Comment:
This stacking of environments could use some comments. I am assuming that this is mainly used for function calls, so environments do not have access to each other, right?

Reply:
yep!
- [ ] document stacked environments